osquery/tables/applications/darwin/browser_plugins.cpp

```diff
@@ -152,45 +147,59 @@ inline void genSafariExtension(const std::string& uid,
     return;
   }
 
-  // Use open_file, instead of the preferred open_filename for OS X 10.9.
-  archive_read_support_format_xar(ext);
-  if (archive_read_open_filename(ext, path.c_str(), 10240) != ARCHIVE_OK) {
-    archive_read_finish(ext);
+  xar_t xar = xar_open(path.c_str(), READ);
+  if (xar == nullptr) {
+    TLOG << "Cannot open extension archive: " << path;
     return;
   }
 
-  struct archive_entry* entry = nullptr;
-  while (archive_read_next_header(ext, &entry) == ARCHIVE_OK) {
-    auto item_path = archive_entry_pathname(entry);
-    // Documentation for libarchive mentions these APIs may return NULL.
-    if (item_path == nullptr) {
-      archive_read_data_skip(ext);
-      continue;
+  xar_iter_t iter = xar_iter_new();
+  xar_file_t xfile = xar_file_first(xar, iter);
+
+  size_t max_files = 500;
+  for (size_t index = 0; index < max_files; ++index) {
+    if (xfile == nullptr) {
+      break;
     }
 
-    // Assume there is no non-root Info.
-    if (std::string(item_path).find("Info.plist") == std::string::npos) {
-      archive_read_data_skip(ext);
-      continue;
+    char* xfile_path = xar_get_path(xfile);
+    if (xfile_path == nullptr) {
+      break;
     }
 
-    // Read the decompressed Info.plist content.
-    auto content = std::string(archive_entry_size(entry), '\0');
-    archive_read_data_into_buffer(ext, &content[0], content.size());
+    // Clean up the allocated content ASAP.
+    std::string entry_path(xfile_path);
+    free(xfile_path);
+    if (entry_path.find("Info.plist") != std::string::npos) {
+      if (xar_verify(xar, xfile) != XAR_STREAM_OK) {
+        TLOG << "Extension info extraction failed verification: " << path;
+      }
+
+      size_t size = 0;
+      char* buffer = nullptr;
+      if (xar_extract_tobuffersz(xar, xfile, &buffer, &size) != 0 ||
```

Comment:
I don't trust these APIs, this should include a final check for an allocated `buffer` as to not subtly leak.